### PR TITLE
[HWASAN] Implement selective instrumentation based on profiling information

### PR DESCRIPTION
**Pull request by kstoimenov**

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 1]
```diff
@@ -1501,11 +1520,39 @@ bool HWAddressSanitizer::instrumentStack(memtag::StackInfo &SInfo,
...
+    NumNoSanitizeFuncs++;
     return;
+  }
+
+  NumConsideredFuncs++;
+  if (CSkipHotCode) {
+    auto &MAMProxy = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
+    ProfileSummaryInfo *PSI =
+        MAMProxy.getCachedResult<ProfileSummaryAnalysis>(*F.getParent());
+    if (PSI != nullptr && PSI->hasProfileSummary()) {
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1539]: just `PSI &&`

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 2]
```diff
@@ -1501,11 +1520,39 @@ bool HWAddressSanitizer::instrumentStack(memtag::StackInfo &SInfo,
...
+  if (&F == HwasanCtorFunction) {
+    NumHwasanCtors++;
     return;
+  }
 
-  if (!F.hasFnAttribute(Attribute::SanitizeHWAddress))
+  if (!F.hasFnAttribute(Attribute::SanitizeHWAddress)) {
+    NumNoSanitizeFuncs++;
     return;
+  }
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1532]: Also:
```
if (F.empty())
  return;
```

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 3]
```diff
@@ -1501,11 +1520,39 @@ bool HWAddressSanitizer::instrumentStack(memtag::StackInfo &SInfo,
...
+
+      if (is_hot) {
+        ++NumSkippedHotFuncs;
+        return;
+      }
+    } else {
+      ++NumNoProfileSummaryFuncs;
+    }
+  }
+  NumInstrumentedFuncs++;
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1555]: Let's not spam stats:
`NumInstrumentedFuncs = NumConsideredFuncs - NumSkippedHotFuncs;`

NumConsideredFuncs -> NumTotalFuncs?

And I don't see what how we will use NumHwasanCtors and NumNoSanitizeFuncs, maybe remove?

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 4]
```diff
@@ -1501,11 +1520,39 @@ bool HWAddressSanitizer::instrumentStack(memtag::StackInfo &SInfo,
...
+    auto &MAMProxy = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
+    ProfileSummaryInfo *PSI =
+        MAMProxy.getCachedResult<ProfileSummaryAnalysis>(*F.getParent());
+    if (PSI != nullptr && PSI->hasProfileSummary()) {
+      auto &BFI = FAM.getResult<BlockFrequencyAnalysis>(F);
+      const bool is_hot =
+          (HotPercentileCutoff.getNumOccurrences() && HotPercentileCutoff >= 0)
+              ? PSI->isFunctionHotInCallGraphNthPercentile(HotPercentileCutoff,
+                                                           &F, BFI)
+              : PSI->isFunctionEntryHot(&F);
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1545]: isFunctionEntryHot -> isFunctionHotInCallGraph

File: llvm/test/Instrumentation/HWAddressSanitizer/selsan-no-ps.ll [hunk 5]
```diff
@@ -0,0 +1,16 @@
+; RUN: opt < %s -passes='require<profile-summary>,hwasan' -S -stats 2>&1 \
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/HWAddressSanitizer/selsan-no-ps.ll:1]: selsan -> `pgo-opt-out`

**Review by vitalybuka — APPROVED** [review]

**Comment by Enna1** [comment]
Hi @kstoimenov @vitalybuka 

I have some high level questions:
1. What is the motivation of profile information based selective instrumentation? is it about code size?
2. Will the profile information based selective instrumentation be implemented for other sanitizers(asan, msan, tsan)?
3. Will the profile information based selective instrumentation cause some false negatives or false positives?
IIUR, lets say the compilation unit where the heap-buffer-overflow happens is not instrumented, then there is no hwasan check is inserted, so false negative may happen?

Thanks!

**Comment by RKSimon** [comment]
@kstoimenov  @vitalybuka  This appears to be breaking some bots: 

https://lab.llvm.org/buildbot/#/builders/124/builds/9846
Failed Tests (5):
  LLVM :: CodeGen/AArch64/shufflevector.ll
  LLVM :: CodeGen/Hexagon/loop_align_count.ll
  LLVM :: CodeGen/Hexagon/loop_align_count.mir
  LLVM :: Instrumentation/HWAddressSanitizer/pgo-opt-out-no-ps.ll
  LLVM :: Instrumentation/HWAddressSanitizer/pgo-opt-out.ll

**Comment by kstoimenov** [comment]
I reverted it with f6f79d46e580b34b2c98bd9bda7ede3a38f43f77.

**Comment by kstoimenov** [comment]
Reapplied in 56754478e7184a68e88ccc578ffd31097da85d15.

**Comment by vitalybuka** [comment]
> Hi @kstoimenov @vitalybuka
> 
> I have some high level questions:
> 
> 1. What is the motivation of profile information based selective instrumentation? is it about code size?

There are use cases when we can't run 100% of instrumentation because of unacceptable performance overhead. However if we opt out hot code we can still keep most of coverage with minimal loss.

> 2. Will the profile information based selective instrumentation be implemented for other sanitizers(asan, msan, tsan)?

ubsan, hwasan only.

To avoid false positives msan, tsan even with no_sanitize attribute require some instrumentation, so perf overhead will be smaller, but still significant. So I think it's infeasible here.

asan for good detection requires allocator quarantine and fake stack. So it may benefit, but still expensive. We don't have plans on that, but anyone is welcome to try to port this patch to Asan.

> 3. Will the profile information based selective instrumentation cause some false negatives or false positives?
>    IIUR, lets say the compilation unit where the heap-buffer-overflow happens is not instrumented, then there is no hwasan check inserted, so false negative may happen?

non-instrumented code with hwasan, like with asan, will not cause false positive.
Some false negative are expected by definition, if code involved into the bug is hot and instrumentation is omitted.

> 
> Thanks!
> 
> edited: Just noticed this ubsan patch: #83471

**Pull request merged**